### PR TITLE
Fix for CondFD Problem that Improperly Led to Fatal Error 

### LINKED ISSUE #9597: SurfaceProperty:HeatTransferAlgorithm:Construction leads to erroneous fatal error InitialInitHeatBalFiniteDiff
Issue overview
--------------
Using SurfaceProperty:HeatTransferAlgorithm:Construction leads to erroneous fatal error InitialInitHeatBalFiniteDiff reporting problems with a material not part of the construction which is addressed in the SurfaceProperty block.

### Details
Some additional details for this issue (if relevant):
 - MacOS, 12.4
 - EnergyPlus, Version 22.2.0-30ad94ee2a, YMD=2022.08.11 14:07
 - https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Funmethours.com%2Fquestion%2F77266%2Fenergyplus-setting-heat-transfer-algorithm-for-constructions-leads-to-fatal-error%2F&data=05%7C01%7Cachim.geissler%40fhnw.ch%7C13f22b975f0d4596f83208da6fdf5e20%7C9d1a5fc8321e4101ae63530730711ac2%7C0%7C0%7C637945302393052061%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=zfr6CfK632b5YwgWU55b3esZ1B76%2FlGlPDOpK4NHGo0%3D&reserved=0 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x ] "Defect" file added (https://drive.switch.ch/index.php/s/6pVASJKy0nGnvM7/download)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #9597
 - User had an input file that requested CondFD for a single construction that was leading to a fatal error crash.  The problem was caused by an input file that used regular materials to define air gaps and then asked for one construction to use the CondFD solution (that didn't actually use that regular material defined air gap).  The CondFD code was set up to establish arrays for all constructions except a couple scenarios.  This caused it to try to initialize data for a construction that was not even using the CondFD solution and led to problems in the initialization routine (i.e., a fatal error to avoid a potentially unstable situation).  The problem went away when the material was defined using Material:AirGap because the code had logic for that scenario while is entirely appropriate.  However, there was no reason for the CondFD algorithm to go constructions that did not use the CondFD solution.  In most cases, it didn't cause a problem.  In this case, it threw an inappropriate fatal error.  The fix involves a new subroutine that checks to see if a particular construction actually uses CondFD and skipping the ones that don't.  This allowed the user's input file that caused the problem in the first place to run to completion as expected.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HeatBalFiniteDiffManager.cc [hunk 1]
```diff
@@ -2626,6 +2629,17 @@ namespace HeatBalFiniteDiffManager {
         updatedThermalConductivity = materialDefinition.phaseChange->getConductivity(temperatureUpdated);
     }
 
+    bool findAnySurfacesUsingConstructionAndCondFD(EnergyPlusData &state, int constructionNum)
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalFiniteDiffManager.cc]: int const contructionNum, since it's not changing.

Comment (RKStrand) [reply src/EnergyPlus/HeatBalFiniteDiffManager.cc]: Yup.  This was an oversight on my part.  I've fixed it in the latest commit.

File: src/EnergyPlus/HeatBalFiniteDiffManager.cc [hunk 2]
```diff
@@ -2626,6 +2629,17 @@ namespace HeatBalFiniteDiffManager {
...
     }
 
+    bool findAnySurfacesUsingConstructionAndCondFD(EnergyPlusData &state, int const constructionNum)
+    {
+        for (int surfNum = 1; surfNum <= state.dataSurface->TotSurfaces; ++surfNum) {
+            auto &thisSurface = state.dataSurface->Surface(surfNum);
+            if (thisSurface.Construction == constructionNum) {
+                if (thisSurface.HeatTransferAlgorithm == DataSurfaces::HeatTransferModel::CondFD) return true;
+            }
+        }
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalFiniteDiffManager.cc:2638]: for (auto &thisSurface : state.dataSurface->Surface) {
            if (thisSurface.Construction == constructionNum) {
                if (thisSurface.HeatTransferAlgorithm == DataSurfaces::HeatTransferModel::CondFD) return true;
            }
        }

Update: use auto &

**Comment by rraustad** [comment]
@RKStrand the defect file AUE_BS from issue 9597 runs on Windows in develop. I added Output:Constructions,Contructions,Materials; to the defect file and the file still runs in develop. Is there something else that needs to be added to the defect file? I am trying to hit your new function in this branch in the debugger (to verify my suggested code change) and have yet to do so.

**Comment by RKStrand** [comment]
@rraustad I think that the unmet hours post clarifies what change you have to make to get the problem.  One of the surfaces in the file has to be changed to use CondFD.  The file itself uses CTFs but when you switch one surface to CondFD, you should see the problem happen.

File: src/EnergyPlus/HeatBalFiniteDiffManager.cc (hunk 2 above)
Comment (RKStrand) [reply src/EnergyPlus/HeatBalFiniteDiffManager.cc:2638]: Ok--made this change as suggested and pushed the commit.

**Comment by rraustad** [comment]
OK, the defect file was changed as:

    !-   ===========  ALL OBJECTS IN CLASS: HEATBALANCEALGORITHM ===========
    HeatBalanceAlgorithm,ConductionTransferFunction,200;

    SurfaceProperty:HeatTransferAlgorithm:Construction,
      BO05Override,            !- Name
      ConductionFiniteDifference,  !- Algorithm          <-- changed in defect file
      BO05;                    !- Construction Name

    SurfaceProperty:HeatTransferAlgorithm:Construction,
      BO05ReversedOverride,    !- Name
      ConductionFiniteDifference,  !- Algorithm         <-- changed in defect file
      BO05 Reversed;           !- Construction Name


Now this defect file is showing fatal errors in develop for material layers. Changing the material thickness as the message suggests leads to another material layer with this same fatal warning.

    ** Severe  ** InitialInitHeatBalFiniteDiff: Found Material that is too thin and/or too highly conductive, material name = LUFTSCHICHT_9CM
    **   ~~~   ** High conductivity Material layers are not well supported by Conduction Finite Difference, material conductivity = 0.630 [W/m-K]
    **   ~~~   ** Material thermal diffusivity = 5.122E-004 [m2/s]
    **   ~~~   ** Material with this thermal diffusivity should have thickness > 0.52591 [m]
    **  Fatal  ** Preceding conditions cause termination.

This branch runs the defect file successfully.

    ************* EnergyPlus Completed Successfully-- 124 Warning; 0 Severe Errors; Elapsed Time=00hr 16min 11.21sec

**Comment by rraustad** [comment]
Review is complete, just waiting on CI to finish and check for diffs and performance impact. This should merge shortly.

**Comment by rraustad** [comment]
There is 1 eio diff in 1ZoneUncontrolledFourAlgorithms that needs to be explained. This diff is basically saying that there is information that is now not being reported. @RKStrand can you explain that?

image

**Comment by rraustad** [comment]
CI is clean with a single time out on DirectIndirectEvapCoolersVSAS.

**Comment by RKStrand** [comment]
@rraustad Yes, this is actually correct because in develop it forces CondFD to do calculates for every surface even if only one uses the CondFD solution.  As a result, it produces a bunch of output that is nonsense because it isn't even used.  In this modification, it no longer does those calculations so this junk output does not appear anymore.  So, this is actually correct that it does not produce this output (it's actually an additional sign that the fix is doing what it is supposed to).  Does that help?

**Comment by RKStrand** [comment]
I think I set a record for the number of times that the word "actually" was used in a single GitHub comment 🤦‍♂️

**Pull request merged**